cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMetadataStateManager.java

```diff
@@ -0,0 +1,410 @@
+                } else {
+                    // ZooKeeper rolled back on brokers
+                    return KafkaMetadataConfigurationState.ZK;
+                }
+            }
+            case KRaftDualWriting -> {
+                if (isKRaftAnnoDisabled()) {
+                    // ZooKeeper rolled back on brokers
+                    return KafkaMetadataConfigurationState.ZK;
+                } else if (isKRaftAnnoMigration()) {
+                    // ZooKeeper still configured on controllers, removed on brokers
+                    return KafkaMetadataConfigurationState.POST_MIGRATION;
+                } else if (isKRaftAnnoRollback()) {
+                    // a rollback is going on, back in dual writing with ZooKeeper configured on both controllers and brokers
+                    return KafkaMetadataConfigurationState.MIGRATION;
+                }
+            }
+            case KRaftPostMigration -> {
+                if (isKRaftAnnoEnabled()) {
+                    return KafkaMetadataConfigurationState.KRAFT;
+                // rollback
+                } else if (isKRaftAnnoRollback()) {
+                    // ZooKeeper configured and migration enabled on both controllers and brokers
+                    return KafkaMetadataConfigurationState.MIGRATION;
+                } else {
+                    // ZooKeeper still configured on controllers, removed on brokers
+                    return KafkaMetadataConfigurationState.POST_MIGRATION;
+                }
+            }
+            case PreKRaft, KRaft -> {
+                return KafkaMetadataConfigurationState.KRAFT;
+            }
+        }
+        // this should never happen
+        throw new IllegalArgumentException("Invalid internal Kafka metadata state [" + this.metadataState + "] with strimzi.io/kraft annotation [" + this.kraftAnno + "]");
+    }
+
+    /**
+     * Check for the status of the Kafka metadata migration
+     *
+     * @param reconciliation    Reconciliation information
+     * @param clusterCaCertSecret   Secret with the Cluster CA public key
+     * @param coKeySecret   Secret with the Cluster CA private key
+     * @param controllerPodName Name of the quorum controller leader pod
+     */
+    public void checkMigrationInProgress(Reconciliation reconciliation, Secret clusterCaCertSecret, Secret coKeySecret, String controllerPodName) {
+        KafkaAgentClient kafkaAgentClient = initKafkaAgentClient(reconciliation, clusterCaCertSecret, coKeySecret);
+        KRaftMigrationState kraftMigrationState = kafkaAgentClient.getKRaftMigrationState(controllerPodName);
+        LOGGER.infoCr(reconciliation, "ZooKeeper to KRaft migration state {} checked on controller {}", kraftMigrationState.state(), controllerPodName);
+        if (kraftMigrationState.state() == -1) {
+            throw new RuntimeException("Failed to get the ZooKeeper to KRaft migration state");
+        }
+        this.isMigrationDone = kraftMigrationState.isMigrationDone();
+    }
```

Comment:
I would probably design this differently:
* Have a static method `KRaftMigrationUtils` to get the migration state from the Kafka nodes (see also similar comment in `KafkaRecocniler`)
* In `KafkaMetadataStateManager` have only the management of the states and some method you would call to transition the state based on the result of the Kafka agent check.

I think that would provide much better separation (this class would not need to know anything about Kafka Agent clients, and just handle the state) and make it much easier to test. It would make it also much more transparent of that is happening in `KafkaReconciler`. And surely it would be much easier to test because ou can independently test how you get the state from the Kafka agents and what does the state manager  do based on different outcomes of that.

Reply:
I have refactored this way.